cc3d/core/PyCoreSpecs.py

```diff
@@ -5309,6 +5379,114 @@ def from_xml(cls, _xml: CC3DXMLElement):
         return cls()
 
 
+class SurfaceTrackerPlugin(_PyCorePluginSpecs, _PyCoreSteerableInterface):
+    """ SurfaceTracker Plugin """
+
+    name = "surface_tracker"
+    registered_name = "SurfaceTracker"
+
+    def __init__(self,
+                 neighbor_order: int = None,
+                 max_neighbor_order: int = None,
+                 max_neighbor_distance: float = None):
+
+        super().__init__()
+
+        self.max_neighbor_distance = max_neighbor_distance
+        self.max_neighbor_order = max_neighbor_order
+        self.neighbor_order = neighbor_order
+
+    @property
+    def xml(self) -> ElementCC3D:
+        """
+        CC3DXML element; generated from specification dictionary
+
+        :return: CC3DML XML element
+        """
+        self._el = self.generate_header()
+        if self.neighbor_order is not None:
+            self._el.ElementCC3D('NeighborOrder', {}, str(self.neighbor_order))
+        elif self.max_neighbor_order is not None:
+            self._el.ElementCC3D('MaxNeighborOrder', {}, str(self.max_neighbor_order))
+        elif self.max_neighbor_distance is not None:
+            self._el.ElementCC3D('MaxNeighborDistance', {}, str(self.max_neighbor_distance))
+        return self._el
+
+    @classmethod
+    def from_xml(cls, _xml: CC3DXMLElement):
+        """
+        Instantiate an instance from a CC3DXMLElement parent instance
+
+        :param _xml: parent xml
+        :return: python class instace
+        :rtype: MomentOfInertiaPlugin
+        """
+        el = cls.find_xml_by_attr(_xml)
+
+        neighbor_order = None
+        max_neighbor_order = None
+        max_neighbor_distance = None
+        if el.findElement('NeighborOrder'):
+            neighbor_order = el.getFirstElement('NeighborOrder').getUInt()
+        elif el.findElement('MaxNeighborOrder'):
+            max_neighbor_order = el.getFirstElement('MaxNeighborOrder').getUInt()
+        elif el.findElement('MaxNeighborDistance'):
+            max_neighbor_distance = el.getFirstElement('MaxNeighborDistance').getDouble()
+
+        return cls(neighbor_order, max_neighbor_order, max_neighbor_distance)
+
+    @property
+    def neighbor_order(self) -> Optional[int]:
+        return self.spec_dict['neighbor_order']
+
+    @neighbor_order.setter
+    def neighbor_order(self, _val: Optional[int]):
+        if _val is None:
+            val = _val
+        else:
+            val = int(_val)
+            if val <= 0:
+                raise ValueError('Value must be positive')
+            self.max_neighbor_order = None
+            self.max_neighbor_distance = None
```

Comment:
I am not sure about having these side effects wrapped in the setters. Is it normal for this plugin?

Reply:
This is imposing object state consistently with how the XML spec is parsed and implemented, to avoid confusion for the user. For example, if two underlying attributes are not `None`, the XML will only use one of them, so the object states reflect this.

Reply:
That makes sense. I like that design, then.